apiserver/uniter/subordinaterelationwatcher.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package uniter
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/utils/set"
+	"gopkg.in/juju/charm.v6-unstable"
+
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/watcher"
+	"github.com/juju/juju/worker/catacomb"
+)
+
+type subRelationsWatcher struct {
+	catacomb      catacomb.Catacomb
+	backend       *state.State
+	app           *state.Application
+	principalName string
+
+	// Maps relation keys to whether that relation should be
+	// included. Needed particularly for when the relation goes away.
+	relations map[string]bool
+	out       chan []string
+}
+
+// newSubordinateRelationsWatcher creates a watcher that will notify
+// about relation lifecycle events for subordinateApp, but filtered to
+// be relevant to a unit deployed to a container with the
+// principalName app. Global relations will be included, but only
+// container-scoped relations for the principal application will be
+// emitted - other container-scoped relations will be filtered out.
+func newSubordinateRelationsWatcher(backend *state.State, subordinateApp *state.Application, principalName string) (
```

Comment:
I implemented this before finding out that the LifecycleWatcher can take a filter function to do this kind of thing. Let me know if this needs to be rewritten using that.